MinIO VMware Documentation WIP (#302)
minio/docs: source/vmware/vmware-overview.rst

.. _minio-vmware:

=======================
MinIO on VMware vSphere
=======================

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 2

Overview
--------

MinIO is a Kubernetes-native, high performance object storage suite that
maintains AWS S3 API compatibility, allowing applications to seamlessly migrate
to private or hybrid cloud infrastructures.  

> Starting with VMware vSphere 7 Update 1 (7U1), vSphere includes MinIO's
> first-party integration through the vSAN Data Persistence Platform (DPP). 
> IT administrators using vSphere can provision high-performance multi-tenant
> object storage entirely through the vCenter interface and with minimal
> manual interaction with the underlying Kubernetes infrastructure. 

Reviewer: ```suggestion
Starting with VMware vSphere 7 Update 1 (7U1), vSphere includes MinIO's
first-party integration through the vSAN Data Persistence Platform (DPP). 
IT administrators using vSphere can provision high-performance multi-tenant
object storage entirely through the vCenter interface. 
```

two vSAN data paths specifically designed for cloud-native stateful
services like MinIO.

.. class:: bold-dt

vSAN SNA Drives
   Allows creating vSAN storage objects that are co-located with their
   associated compute resources. Each MinIO pod only uses drives that are local
   to the node it is running on. The pod relies on MinIO :ref:`Erasure Coding
   <minio-erasure-coding>` for durability or availability of data and
   makes no use of vSAN durability or availability features.

vSAN Direct
   Allows creating independent datastores on every disk attached to a node,
   where applications running on that node can have direct access to those
   disks. The MinIO Tenant can fully utilize the available throughput of the
   locally attached disk without the overhead of network-attached storage.

   vSAN Direct drives are ideal for running MinIO Tenants.

While MinIO can make efficient use of either data path, vSAN Direct is aligned
with MinIO's best practices around ensuring a fast path between a MinIO process
and its storage. vSAN Direct also allows for independent fault domains for the
disks used by a MinIO Pod.

> MinIO recommends using only *one* type of vSAN data path for all drives intended
> for use by a given MinIO Tenant. vSphere administrators should create
> a :vmware7-docs:`Storage Policy
> <com.vmware.vsphere.storage.doc/GUID-720298C6-ED3A-4E80-87E8-076FFF02655A.html>`
> associated to either the vSAN SNA or vSAN Direct disks and ensure that the
> :vmware7-docs:`namespace
> <vmware-vsphere-with-tanzu/GUID-1544C9FE-0B23-434E-B823-C59EFC2F7309.html>` in
> which the MinIO Tenant will be deployed has access to that policy.

Reviewer: ```suggestion
MinIO only uses *one* type of vSAN data path for all drives intended
for use by a given MinIO Tenant. vSphere administrators should create
a :vmware7-docs:`Storage Policy
<com.vmware.vsphere.storage.doc/GUID-720298C6-ED3A-4E80-87E8-076FFF02655A.html>`
associated to either the vSAN SNA or vSAN Direct disks and ensure that the
:vmware7-docs:`namespace
<vmware-vsphere-with-tanzu/GUID-1544C9FE-0B23-434E-B823-C59EFC2F7309.html>` in
which the MinIO Tenant will be deployed has access to that policy.
```

.. _minio-vmware:

=======================

> MinIO on VMware vSphere

Author: @kerneltime so this is a rough initial draft, but I'd like to get your opinion on how this reads.

In particular:

1) Does this accurately describe our relationship with VMware vSphere components
2) Does this accurately describe how MinIO works on vSphere

Thanks.

Author: @dvaldivia if you want to squeeze in a looksie before you go on vacation I wont stop you :)

=======================

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 2

Overview
--------

> MinIO is a Kubernetes-native, high performance object storage suite that
> maintains AWS S3 API compatibility, allowing applications to seamlessly migrate
> to private or hybrid cloud infrastructures. VMware Cloud Foundation (VCF) with
> Tanzu is an enterprise-grade platform for running both traditional and modern
> Kubernetes-based applications in production with all the power of virtualized
> storage, networking, and associated management. 

Reviewer: Customers visiting this place would have already decide how their VMware on prem looks like and if we include this then our docs will have to be in sync with the various ways VMware can choose productize their vSAN DPP offering
```suggestion
MinIO is a Kubernetes-native, high performance object storage suite that
maintains AWS S3 API compatibility, allowing applications to seamlessly migrate
to private or hybrid cloud infrastructures.  
```

Starting with VMware vSphere 7 Update 1 (7U1), vSphere includes MinIO's
first-party integration through the vSAN Data Persistence Platform (DPP). 
IT administrators using vSphere can provision high-performance multi-tenant
object storage entirely through the vCenter interface and with minimal
manual interaction with the underlying Kubernetes infrastructure. 

IMAGE

This documentation assumes familiarity with all referenced VMware components,
including but not limited to VMware vCenter,VMware vSAN, VMware vSAN Data

> Persistence Platform, vSAN Direct Configuration, and vSAN SNA. While this

Reviewer: ```suggestion
Persistence Platform, vSAN Direct, and vSAN SNA. While this
```

documentation *may* provide guidance for configuring or deploying VMware
components on a best-effort basis, it is not a replacement for dedicated
VMware-specific documentation.

vSAN Direct and vSAN SNA Storage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

VMware vSphere 7U1 and the VMware Data Persistence Platform (DPP) offer

> two new vSAN data paths specifically designed for cloud-native stateful

Reviewer: ```suggestion
two vSAN data paths specifically designed for cloud-native stateful
```

services like MinIO.

.. class:: bold-dt

vSAN SNA Drives

>    Allows creating vSAN storage objects that are co-located with their
>    associated compute resources. The MinIO tenant can utilize the local storage
>    with less overhead as compared to traditional vSAN networked storage.

Reviewer: ```suggestion
   Allows creating vSAN storage objects that are co-located with their
   associated compute resources. A MinIO pod will only use drives that are local the node it is running on and will not rely on vSAN for durability or availability of data but provide it using MinIO's erasure coding.
```

>    vSAN SNA drives are ideal for clusters where the MinIO Tenant must share
>    the physical infrastructure with other stateful or stateless workloads.

Reviewer: ```suggestion
```

> vSAN Direct Drives

Reviewer: ```suggestion
vSAN Direct
```

   Allows creating independent datastores on every disk attached to a node,
   where applications running on that node can have direct access to those
   disks. The MinIO Tenant can fully utilize the available throughput of the

>    drive with none of the overhead associated to distributed storage.

Reviewer: ```suggestion
   drive without network overhead, as expected from a local attached disk.
```

>    vSAN Direct drives are ideal for clusters dedicated to running the MinIO
>    Tenant.

Reviewer: ```suggestion
   vSAN Direct drives are ideal for running the MinIO Tenant.
```

> While MinIO can make efficient use of either data path, vSAN Direct most closely
> mirrors MinIO best practices around ensuring the shortest possible path between
> a MinIO process and its connected storage.

Reviewer: ```suggestion
While MinIO can make efficient use of either data path, vSAN Direct is well aligned with MinIO's best practices around ensuring a fast path between
a MinIO process and its storage. vSAN Direct also allows for independent fault domains for the disks used by a MinIO Pod.
```

MinIO recommends using only *one* type of vSAN data path for all drives intended
for use by a given MinIO Tenant. vSphere administrators should create
a :vmware7-docs:`Storage Policy
<com.vmware.vsphere.storage.doc/GUID-720298C6-ED3A-4E80-87E8-076FFF02655A.html>`
associated to either the vSAN SNA or vSAN Direct disks and ensure that the
:vmware7-docs:`namespace
<vmware-vsphere-with-tanzu/GUID-1544C9FE-0B23-434E-B823-C59EFC2F7309.html>` in
which the MinIO Tenant will be deployed has access to that policy.

.. ToDo: Link to vmware docs on configuring/provisioning Direct/SNA Drives

.. _minio-vmware-enable:

Enable MinIO for VMware vSphere
-------------------------------

> VMware vSphere 7 Update 1 includes the MinIO Client Plugin. You have to

Reviewer: Update 1 is already out.. we need to find the exact patch that will include MinIO
```suggestion
VMware vSphere 7 Update 1 XXX includes the MinIO Client Plugin. You have to
```

plugin. 

IMAGE

Use the :ref:`minio-vmware-getting-started` procedure to deploy a MinIO Tenant
for evaluation and development. See :ref:`minio-vmware-create-tenant` for a
complete guide to deploying MinIO Tenants using vCenter.

.. _minio-vmware-getting-started:

Getting Started
---------------

The following procedure creates a MinIO Tenant through the VMware vCenter
interface. This procedure does not cover advanced configuration such as
Server Side Encryption (SSE-S3) or external IDentity Providers (IDP). 
See <Future Doc> for a complete description of all Tenant creation options.

Prerequisites
~~~~~~~~~~~~~

This procedure assumes the following:

.. list-table::
   :header-rows: 1
   :widths: 30 70
   :width: 100%

   * - Component
     - 

   * - VMware vSphere v7 Update 1
     - 

   * - VMware ESXi Hosts
     - Ensure the cluster has *at least* as many ESXi hosts as MinIO nodes
       which you want to deploy.

   * - :vmware7-docs:`Storage Policy 
       <com.vmware.vsphere.storage.doc/GUID-720298C6-ED3A-4E80-87E8-076FFF02655A.html>` 
     - MinIO ensures any PVCs generated to support the Tenant bind *only* to
       PVs associated to the specified storage policy. 

>        The Storage Policy should point to a group of *either*:

Reviewer: Storage policy is just a policy, this can reworded as Select either vSAN Direct or vSAN SNA based storage policy.

       - vSAN Direct disks *or*

       - vSAN SNA disks

>        The disks *must* be directly connected to the ESXi hosts onto which MinIO
>        deploys. MinIO strongly recommends *against* provisioning networked
>        storage for use with the Tenant.

Reviewer: ```suggestion
vSAN Direct and vSAN SNA allow for local attached disks. MinIO strongly recommends against networked storage.
```

Reviewer: `enable host local disks only when using vSAN`?

   * - :vmware7-docs:`Namespace 
       <vmware-vsphere-with-tanzu/GUID-1544C9FE-0B23-434E-B823-C59EFC2F7309.html>`
       

>      - MinIO recommends creating a new namespace for each MinIO Tenant you
>        deploy into the cluster.
> 

Reviewer: ```suggestion
     - MinIO only supports a single tenant within a namespace.

```

Procedure
~~~~~~~~~

1) Navigate to the MinIO Dashboard
``````````````````````````````````

Log in to the vCenter interface and select the Datacenter and Cluster
in which you want to deploy the MinIO Tenant:

IMAGE

Select :guilabel:`Configure` from the navigation bar and scroll down to the
:guilabel:`MinIO` section.

IMAGE

Click on :guilabel:`MinIO` to expand the section, then click 
:guilabel:`Tenants`.

IMAGE

2) Open the Create Tenant Modal
```````````````````````````````

Click :guilabel:`Add` to open the :guilabel:`Create Tenant` Modal. 
The :guilabel:`Tenant Configuration` step presents the following
configuration fields:

.. list-table::
   :header-rows: 1
   :widths: 20 80
   :width: 100%

   * - Field
     - Description

   * - :guilabel:`Name` 
     - Enter the name of the MinIO Tenant. The specified string acts as a
       prefix for all resources created as part of the Tenant.

   * - :guilabel:`Namespace` 
     - Enter the name of a Namespace in your cluster.
       
       The namespace *must* already exist in the cluster.

   * - :guilabel:`Storage Class`
     - Select an available Storage Class associated to the specified namespace.

>        Ensure the Storage Class provides access to the vSAN Direct Configuration
>        disks in the cluster.

Reviewer: ```suggestion
```
